### ISSUE TITLE
DACE_loss 的实现是否与论文公式（7）存在差异？

### ISSUE THREAD

**Issue by laowang666222** [liang-zibo/DACE#3]
我在阅读代码时发现，model.py 中 DACE_loss 的实现似乎与论文公式（7）的描述存在一些差异，想请教一下。

  根据论文公式（7），每个节点的 loss weight 为 alpha ** height。一个 query plan 的 loss 是将所有节点（包括 root plan 和 sub-plan）的 q-
  error 乘以相应的高度权重后求和。

  但是，目前代码中的实现为：

  def DACE_loss(self, est_run_times, run_times, loss_mask):
      loss = torch.max(
          est_run_times / run_times,
          run_times / est_run_times,
      )
      loss = loss * loss_mask
      loss = torch.log(torch.where(loss > 1, loss, 1))
      loss = torch.sum(loss, dim=1)
      return loss

  这里先计算 qerror * loss_weight，然后再取对数。对于非 padding 节点，实际计算过程相当于：

  max(log(qerror) + log(loss_weight), 0)

  这会产生两个影响：

  1. 只有当 qerror > 1 / loss_weight 时，该节点才会产生 loss 和梯度。例如 alpha=0.5 时，高度为 1、2、3 的节点分别需要 qerror > 2、4、8
     才会参与训练。

  2. 当节点超过上述阈值后，log(loss_weight) 是一个与模型参数无关的常数。因此，此时 loss weight 不会连续地缩放该节点的梯度。它的作用更接
     近一个与节点高度有关的 q-error 阈值，而不是论文中描述的梯度权重。

  请问这里先加权、再取对数是有意设计的吗？

  如果严格按照论文公式（7），是否应该实现为：

  qerror = torch.max(
      est_run_times / run_times,
      run_times / est_run_times,
  )
  loss = torch.sum(qerror * loss_mask, dim=1)

  如果使用 log(qerror) 是出于训练稳定性的考虑，那么是否应该先取对数，再应用高度权重：

  qerror = torch.max(
      est_run_times / run_times,
      run_times / est_run_times,
  )
  loss = torch.sum(torch.log(qerror) * loss_mask, dim=1)

  感谢解答！
